cmd/xl-storage-format-utils.go

```diff
@@ -52,19 +37,34 @@ func getFileInfoVersions(xlMetaBuf []byte, volume, path string) (FileInfoVersion
 		}
 	}
 	fivs.Versions = fivs.Versions[:n]
+	// Update numversions
+	for i := range fivs.Versions {
+		fivs.Versions[i].NumVersions = n
+	}
 	return fivs, nil
 }
 
 func getAllFileInfoVersions(xlMetaBuf []byte, volume, path string) (FileInfoVersions, error) {
 	if isXL2V1Format(xlMetaBuf) {
-		var xlMeta xlMetaV2
-		if err := xlMeta.Load(xlMetaBuf); err != nil {
-			return FileInfoVersions{}, err
+		var versions []FileInfo
+		var err error
+		if buf, _ := isIndexedMetaV2(xlMetaBuf); buf != nil {
+			versions, err = buf.ListVersions(volume, path)
+		} else {
+			var xlMeta xlMetaV2
+			if err := xlMeta.Load(xlMetaBuf); err != nil {
+				return FileInfoVersions{}, err
+			}
+			versions, err = xlMeta.ListVersions(volume, path)
 		}
-		versions, latestModTime, err := xlMeta.ListVersions(volume, path)
 		if err != nil {
 			return FileInfoVersions{}, err
 		}
+
+		var latestModTime time.Time
+		if len(versions) > 0 {
+			latestModTime = versions[0].ModTime
+		}
```

Comment:
```suggestion
		var latestModTime time.Time
		if len(versions) > 0 {
			latestModTime = versions[0].ModTime
		}
```

My suggestion is to return right away like 

```
if len(versions) == 0 {
   return FileInfoVersions{}, nil
}
```

Avoids filling FileInfoVersions() with unexpectedly incomplete data while returning.